src/proxy/connection_manager.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::proxy::error;
+use crate::rbac::Connection;
+use drain;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+pub struct ConnectionDrain {
+    tx: drain::Signal,
+    rx: drain::Watch,
+}
+
+impl ConnectionDrain {
+    fn new() -> Self {
+        let (tx, rx) = drain::channel();
+        ConnectionDrain { tx, rx }
+    }
+}
+
+#[derive(Clone)]
+pub struct ConnectionManager {
+    drains: Arc<RwLock<HashMap<Connection, ConnectionDrain>>>,
+}
+
+impl ConnectionManager {
+    pub fn new() -> Self {
+        ConnectionManager {
+            drains: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub async fn track(self, c: &Connection) -> drain::Watch {
+        match self.drains.read().await.get(c) {
+            Some(cd) => cd.rx.clone(),
+            None => {
+                let cd = ConnectionDrain::new();
+                let rx = cd.rx.clone();
+                //TODO: this was racy, another insert may happen between dropping the read lock and attaining this write lock
+                // try_insert is the best solution once it's no longer a nightly-only experimental API
+                let mut drains = self.drains.write().await;
+                if let Some(w) = drains.get(c) {
+                    return w.rx.clone();
+                }
+                drains.insert(c.clone(), cd);
+                rx
+            }
+        }
+    }
+
+    pub async fn drain(&self, c: &Connection) {
+        match self.drains.clone().write().await.remove(c) {
+            Some(cd) => {
+                cd.tx.drain().await;
```

Comment:
is the lock held here? If so, is drain() possibly blocking?

Reply:
I don't think the write lock should be held here. Remove totally removes a value from the HashMap returning the owned value (not a reference) so there's no writing beyond the single remove and no reference to the lock required. Now you've got me doubting that understanding a bit though 😅

Reply:
It's definitely possible to make that more explicit though if we want.

Reply:
```rust
        let w = self.drains.clone().write().await;
        let cd = w.remove(c);
        drop(w);

        match cd {
            ...
        }
```

Reply:
The lock should be released when the value returned from `write()` is dropped. I am just not sure if its drops when the match body (L67) is done or when the match check (L65) is done

Reply:
```rust
        let cd = self.drains.clone().write().await.remove(c);
        match cd {...}
```

Reply:
probably better break it out like that to align with the standards we have for zt... prioritize understandability for our reviewers over the most idiomatic looking rustlang

Reply:
It turns out the lock is held through the duration of the match. This lead to a deadlock in the track method where the read lock was being held for the entire match and one branch was trying to get a write lock.

Reply:
Use of a central RWLocked connection tracker pretty much means by definition that we will be blocking on add/remove ops, I think.

If that's not acceptable, inverting this via channels is probably the other way.

Reply:
that's fair. we could do some reworking of how policy is applied and retained to ease or eliminate interlocking problems if need be. My thinking was sort of limited to not making those kind changes which maybe is unnecessarily narrow

Reply:
we can also explore using linked lists for less locking

Reply:
[dashmap crate](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html)